mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+                break;
+            }
+
+            let end = start + block_size;
+            if end > self.max_file_size {
+                return Err(RingBufferError::WrapAround.into());
+            }
+
+            let bytes = &self.mmap[start..end];
+            // this is unused memory
+            if bytes.iter().map(|&x| x as usize).sum::<usize>() == 0 {
+                break;
+            }
+
+            let block = binary_deserialize::<BlockHeaderWithHash>(bytes)?;
+            // this means we read bytes that don't make a block, this is
+            // a really bad state to be in as somehow the pointers don't
+            // match to where data really is at.
+            if block.inner().hint() != BLOCK_HINT {
+                return Err(RingBufferError::Validate(BlockError::Hint).into());
+            }
+
+            let inner_block = block.inner();
+            let data_size = inner_block.data_size();
+            let index = inner_block.write_index();
+            start += block_size;
+            let end = start + data_size;
+            let data = load_data(self.mmap, start, data_size, self.max_file_size)?;
+            start = end % self.max_file_size;
+
+            // to move to beginning of next block header
+            // start += 1;
+            validate(&block, &data)?;
+            let publication = binary_deserialize::<Publication>(data.inner())?;
+            let key = Key {
+                offset: index as u64,
+            };
+            vdata.push_back((key, publication));
+        }
+
+        Ok(vdata)
+    }
+
+    fn remove(&mut self, key: usize) -> StorageResult<()> {
+        let timer = Instant::now();
+        let read_index = self.pointers.read;
+        if key != read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::RemovalIndex));
+        }
+        let block_size = serialized_block_size()?;
+
+        let start = key;
+        let mut end = start + block_size;
+        if end > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::WrapAround));
+        }
+
+        let maybe_block = load_block_header(self.mmap, start, block_size, self.max_file_size);
+        if maybe_block.is_err() {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        let mut block = maybe_block.unwrap();
+        if block.inner().hint() != BLOCK_HINT {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        let data_size;
+        {
+            let inner_block = block.inner_mut();
+            inner_block.set_should_not_overwrite(false);
+            data_size = inner_block.data_size();
+        }
+
+        let should_flush = self.should_flush();
+        save_block_header(self.mmap, &block, start, self.max_file_size, should_flush)?;
+
+        end += data_size;
+
+        self.pointers.read = end % self.max_file_size;
+
+        self.flush_state
```

Comment:
I think @vadim-kovalyov has some concerns in another method above. Should we extract common code to a separate method?